From 56d9df3251b6c0e9292e3bf18cc6195bac342b63 Mon Sep 17 00:00:00 2001
From: Jinoh Kang <jinoh.kang.kr@gmail.com>
Date: Thu, 20 Feb 2025 23:40:00 +0900
Subject: [PATCH 298/365] ntdll: Register module dependency for export
 forwarders regardless of whether the dependency is already loaded.

Prepare for generalization to imports from DLLs.

Calling add_module_dependency() multiple times for the same dependency
edge no longer bloats memory usage.
---
 dlls/ntdll/loader.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/dlls/ntdll/loader.c b/dlls/ntdll/loader.c
index 0512c7a5732..23dae40f875 100644
--- a/dlls/ntdll/loader.c
+++ b/dlls/ntdll/loader.c
@@ -987,13 +987,15 @@ static FARPROC find_forwarded_export( HMODULE module, const char *forward, LPCWS
         wm_loaded = TRUE;
     }
 
-    if (wm_loaded)
+    if (wm->ldr.DdagNode != node_ntdll && wm->ldr.DdagNode != node_kernel32)
     {
         if (!imports_fixup_done && importer)
         {
+            /* Prepare for the callee stealing the reference */
+            if (!wm_loaded && wm->ldr.LoadCount != -1) wm->ldr.LoadCount++;
             add_module_dependency( importer->ldr.DdagNode, wm->ldr.DdagNode );
         }
-        else if (process_attach( wm->ldr.DdagNode, NULL ) != STATUS_SUCCESS)
+        else if (wm_loaded && process_attach( wm->ldr.DdagNode, NULL ) != STATUS_SUCCESS)
         {
             ERR( "process_attach failed for forward '%s' used by %s\n",
                  forward, debugstr_w(get_modref( module )->ldr.FullDllName.Buffer) );
-- 
2.50.1

